File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/math/DenseVector.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.ml.math

import breeze.linalg.{SparseVector => BreezeSparseVector, DenseVector => BreezeDenseVector, Vector => BreezeVector}

/**
 * Dense vector implementation of [[Vector]]. The data is represented in a continuous array of
 * doubles.
 *
 * @param data Array of doubles to store the vector elements
 */
case class DenseVector(data: Array[Double]) extends Vector with Serializable {

  /**
   * Number of elements in a vector
   * @return the number of the elements in the vector
   */
  override def size: Int = {
    data.length
  }

  /**
   * Element wise access function
   *
   * @param index index of the accessed element
   * @return element at the given index
   */
  override def apply(index: Int): Double = {
    require(0 <= index && index < data.length, index + " not in [0, " + data.length + ")")
    data(index)
  }

  override def toString: String = {
    s"DenseVector(${data.mkString(", ")})"
  }

  override def equals(obj: Any): Boolean = {
    obj match {
      case dense: DenseVector => data.length == dense.data.length && data.sameElements(dense.data)
      case _ => false
    }
  }

  override def hashCode: Int = {
    java.util.Arrays.hashCode(data)
  }

  /**
   * Copies the vector instance
   *
   * @return Copy of the vector instance
   */
  override def copy: DenseVector = {
    DenseVector(data.clone())
  }

  /** Updates the element at the given index with the provided value
    *
    * @param index Index whose value is updated.
    * @param value The value used to update the index.
    */
  override def update(index: Int, value: Double): Unit = {
    require(0 <= index && index < data.length, index + " not in [0, " + data.length + ")")

    data(index) = value
  }

  /** Returns the dot product of the recipient and the argument
    *
    * @param other a Vector
    * @return a scalar double of dot product
    */
  override def dot(other: Vector): Double = {
    require(size == other.size, "The size of vector must be equal.")

    other match {
      case SparseVector(_, otherIndices, otherData) =>
        otherIndices.zipWithIndex.map {
          case (idx, sparseIdx) => data(idx) * otherData(sparseIdx)
        }.sum
      case _ => (0 until size).map(i => data(i) * other(i)).sum
    }
  }

  /** Returns the outer product (a.k.a. Kronecker product) of `this`
    * with `other`. The result will given in [[org.apache.flink.ml.math.SparseMatrix]]
    * representation if `other` is sparse and as [[org.apache.flink.ml.math.DenseMatrix]] otherwise.
    *
    * @param other a Vector
    * @return the [[org.apache.flink.ml.math.Matrix]] which equals the outer product of `this`
    *         with `other.`
    */
  override def outer(other: Vector): Matrix = {
    val numRows = size
    val numCols = other.size

    other match {
      case sv: SparseVector =>
        val entries = for {
          i <- 0 until numRows
          (j, k) <- sv.indices.zipWithIndex
          value = this(i) * sv.data(k)
          if value != 0
        } yield (i, j, value)

        SparseMatrix.fromCOO(numRows, numCols, entries)
      case _ =>
        val values = for {
          i <- 0 until numRows
          j <- 0 until numCols
        } yield this(i) * other(j)

        DenseMatrix(numRows, numCols, values.toArray)
    }
  }

  /** Magnitude of a vector
    *
    * @return The length of the vector
    */
  override def magnitude: Double = {
    math.sqrt(data.map(x => x * x).sum)
  }

  /** Convert to a [[SparseVector]]
    *
    * @return Creates a SparseVector from the DenseVector
    */
  def toSparseVector: SparseVector = {
    val nonZero = (0 until size).zip(data).filter(_._2 != 0)

    SparseVector.fromCOO(size, nonZero)
  }
}

object DenseVector {

  def apply(values: Double*): DenseVector = {
    new DenseVector(values.toArray)
  }

  def apply(values: Array[Int]): DenseVector = {
    new DenseVector(values.map(_.toDouble))
  }

  def zeros(size: Int): DenseVector = {
    init(size, 0.0)
  }

  def eye(size: Int): DenseVector = {
    init(size, 1.0)
  }

  def init(size: Int, value: Double): DenseVector = {
    new DenseVector(Array.fill(size)(value))
  }

  /** BreezeVectorConverter implementation for [[org.apache.flink.ml.math.DenseVector]]
    *
    * This allows to convert Breeze vectors into [[DenseVector]].
    */
  implicit val denseVectorConverter = new BreezeVectorConverter[DenseVector] {
    override def convert(vector: BreezeVector[Double]): DenseVector = {
      vector match {
        case dense: BreezeDenseVector[Double] => new DenseVector(dense.data)
        case sparse: BreezeSparseVector[Double] => new DenseVector(sparse.toDenseVector.data)
      }
    }
  }
}

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/DenseVector.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.flink.ml.common.matrix;

import org.apache.commons.lang3.StringUtils;

import java.util.Arrays;
import java.util.Random;

/**
 * A dense vector represented by a values array.
 */
public class DenseVector extends Vector {

	/**
	 * The array holding the vector data.
	 */
	private double[] data;

	/**
	 * Create a zero size vector.
	 */
	public DenseVector() {
		this(0);
	}

	/**
	 * Create a size n vector with all elements zero.
	 *
	 * @param n Size of the vector.
	 */
	public DenseVector(int n) {
		this.data = new double[n];
	}

	/**
	 * Create a dense vector with the user provided data.
	 *
	 * @param data The vector data.
	 */
	public DenseVector(double[] data) {
		this.data = data;
	}

	/**
	 * Create a dense vector with all elements one.
	 *
	 * @param n Size of the vector.
	 * @return The newly created dense vector.
	 */
	public static DenseVector ones(int n) {
		DenseVector r = new DenseVector(n);
		Arrays.fill(r.data, 1.0);
		return r;
	}

	/**
	 * Create a dense vector with all elements zero.
	 *
	 * @param n Size of the vector.
	 * @return The newly created dense vector.
	 */
	public static DenseVector zeros(int n) {
		DenseVector r = new DenseVector(n);
		Arrays.fill(r.data, 0.0);
		return r;
	}

	/**
	 * Create a dense vector with random values uniformly distributed in the range of [0.0, 1.0].
	 *
	 * @param n Size of the vector.
	 * @return The newly created dense vector.
	 */
	public static DenseVector rand(int n) {
		Random random = new Random();
		DenseVector v = new DenseVector(n);
		for (int i = 0; i < n; i++) {
			v.data[i] = random.nextDouble();
		}
		return v;
	}

	@Override
	public String serialize() {
		StringBuilder sbd = new StringBuilder();

		for (int i = 0; i < data.length; i++) {
			sbd.append(data[i]);
			if (i < data.length - 1) {
				sbd.append(",");
			}
		}
		return sbd.toString();
	}

	/**
	 * Parse the dense vector from a formatted string.
	 *
	 * @param str A string of comma separated values.
	 * @return The parsed vector.
	 */
	public static DenseVector deserialize(String str) {
		try {
			if (org.apache.flink.util.StringUtils.isNullOrWhitespaceOnly(str)) {
				return new DenseVector();
			}

			int n = StringUtils.countMatches(str, ",") + 1;
			double[] data = new double[n];
			int startPos = 0;
			for (int i = 0; i < n - 1; i++) {
				int commaPos = StringUtils.indexOf(str, ",", startPos);
				data[i] = Double.valueOf(StringUtils.substring(str, startPos, commaPos));
				startPos = commaPos + 1;
			}
			data[n - 1] = Double.valueOf(StringUtils.substring(str, startPos));
			return new DenseVector(data);
		} catch (Exception e) {
			throw new RuntimeException(String.format("Fail to parse vector \"%s\".", str), e);
		}
	}

	/**
	 * y = func(x).
	 */
	public static void apply(DenseVector x, DenseVector y, UnaryOp func) {
		assert (x.data.length == y.data.length);
		for (int i = 0; i < x.data.length; i++) {
			y.data[i] = func.f(x.data[i]);
		}
	}

	/**
	 * y = func(x1, x2).
	 */
	public static void apply(DenseVector x1, DenseVector x2, DenseVector y, BinaryOp func) {
		assert (x1.data.length == y.data.length);
		assert (x2.data.length == y.data.length);
		for (int i = 0; i < y.data.length; i++) {
			y.data[i] = func.f(x1.data[i], x2.data[i]);
		}
	}

	/**
	 * y = func(x, alpha).
	 */
	public static void apply(DenseVector x, double alpha, DenseVector y, BinaryOp func) {
		assert (x.data.length == y.data.length);
		for (int i = 0; i < x.data.length; i++) {
			y.data[i] = func.f(x.data[i], alpha);
		}
	}

	@Override
	public DenseVector clone() {
		return new DenseVector(this.data.clone());
	}

	@Override
	public String toString() {
		return this.serialize();
	}

	@Override
	public int size() {
		return data.length;
	}

	@Override
	public double get(int i) {
		return data[i];
	}

	@Override
	public void set(int i, double d) {
		data[i] = d;
	}

	@Override
	public void add(int i, double d) {
		data[i] += d;
	}

	@Override
	public double normL1() {
		double d = 0;
		for (double t : data) {
			d += Math.abs(t);
		}
		return d;
	}

	@Override
	public double normL2() {
		double d = 0;
		for (double t : data) {
			d += t * t;
		}
		return Math.sqrt(d);
	}

	@Override
	public double normL2Square() {
		double d = 0;
		for (double t : data) {
			d += t * t;
		}
		return d;
	}

	@Override
	public double normInf() {
		double d = 0;
		for (double t : data) {
			d = Math.max(Math.abs(t), d);
		}
		return d;
	}

	@Override
	public DenseVector slice(int[] indices) {
		double[] values = new double[indices.length];
		for (int i = 0; i < indices.length; ++i) {
			if (indices[i] >= data.length) {
				throw new RuntimeException("Index is larger than vector size.");
			}
			values[i] = data[indices[i]];
		}
		return new DenseVector(values);
	}

	@Override
	public DenseVector prefix(double d) {
		double[] data = new double[this.size() + 1];
		data[0] = d;
		System.arraycopy(this.data, 0, data, 1, this.data.length);
		return new DenseVector(data);
	}

	@Override
	public DenseVector append(double d) {
		double[] data = new double[this.size() + 1];
		System.arraycopy(this.data, 0, data, 0, this.data.length);
		data[this.size()] = d;
		return new DenseVector(data);
	}

	@Override
	public void scaleEqual(double d) {
		for (int i = 0; i < this.size(); i++) {
			this.data[i] *= d;
		}
	}

	@Override
	public DenseVector plus(Vector other) {
		if (other instanceof DenseVector) {
			DenseVector r = this.clone();
			DenseVector.apply(this, (DenseVector) other, r, ((a, b) -> a + b));
			return r;
		} else {
			return (DenseVector) other.plus(this);
		}
	}

	@Override
	public DenseVector minus(Vector other) {
		if (other instanceof DenseVector) {
			DenseVector r = this.clone();
			DenseVector.apply(this, (DenseVector) other, r, ((a, b) -> a - b));
			return r;
		} else {
			return (DenseVector) other.scale(-1.0).plus(this);
		}
	}

	@Override
	public DenseVector scale(double d) {
		DenseVector r = this.clone();
		for (int i = 0; i < this.size(); i++) {
			r.data[i] *= d;
		}
		return r;
	}

	/**
	 * Set the data of the vector the same as those of another vector.
	 */
	public void setEqual(DenseVector other) {
		System.arraycopy(other.data, 0, this.data, 0, this.size());
	}

	/**
	 * this += other .
	 */
	public void plusEqual(DenseVector other) {
		DenseVector.apply(this, other, this, ((a, b) -> a + b));
	}

	/**
	 * this += alpha * other .
	 */
	public void plusScaleEqual(DenseVector other, double alpha) {
		for (int i = 0; i < this.size(); i++) {
			this.data[i] += other.data[i] * alpha;
		}
	}

	/**
	 * this -= other .
	 */
	public void minusEqual(DenseVector other) {
		DenseVector.apply(this, other, this, ((a, b) -> a - b));
	}

	/**
	 * this -= alpha * other .
	 */
	public void minusScaleEqual(DenseVector other, double alpha) {
		for (int i = 0; i < this.size(); i++) {
			this.data[i] -= other.data[i] * alpha;
		}
	}

	/**
	 * this = max(this, other) .
	 */
	public void maxEqual(DenseVector other) {
		DenseVector.apply(this, other, this, ((a, b) -> Math.max(a, b)));
	}

	/**
	 * this = min(this, other) .
	 */
	public void minEqual(DenseVector other) {
		DenseVector.apply(this, other, this, ((a, b) -> Math.min(a, b)));
	}

	@Override
	public double dot(Vector vec) {
		if (vec instanceof DenseVector) {
			DenseVector other = (DenseVector) vec;
			double d = 0;
			for (int i = 0; i < this.size(); i++) {
				d += this.data[i] * other.data[i];
			}
			return d;
		} else {
			return ((SparseVector) vec).dot(this);
		}
	}

	/**
	 * Round each elements and create a new vector to store the result.
	 *
	 * @return The newly created dense vector.
	 */
	public DenseVector round() {
		DenseVector r = new DenseVector(this.size());
		for (int i = 0; i < this.size(); i++) {
			r.data[i] = Math.round(this.data[i]);
		}
		return r;
	}

	/**
	 * Round each elements inplace.
	 */
	public void roundEqual() {
		for (int i = 0; i < this.size(); i++) {
			this.data[i] = Math.round(this.data[i]);
		}
	}

	/**
	 * Compute the outer product with itself.
	 *
	 * @return The outer product matrix.
	 */
	public DenseMatrix outer() {
		return this.outer(this);
	}

	/**
	 * Compute the outer product with another vector.
	 *
	 * @return The outer product matrix.
	 */
	public DenseMatrix outer(DenseVector other) {
		int nrows = this.size();
		int ncols = other.size();
		double[] data = new double[nrows * ncols];
		int pos = 0;
		for (int j = 0; j < ncols; j++) {
			for (int i = 0; i < nrows; i++) {
				data[pos++] = this.data[i] * other.data[j];
			}
		}
		return new DenseMatrix(nrows, ncols, data, false);
	}

	/**
	 * Get the data array.
	 */
	public double[] getData() {
		return this.data;
	}

	/**
	 * Set the data array.
	 */
	public void setData(double[] data) {
		this.data = data;
	}

	@Override
	public void standardizeEqual(double mean, double stdvar) {
		int size = data.length;
		for (int i = 0; i < size; i++) {
			data[i] -= mean;
			data[i] *= (1.0 / stdvar);
		}
	}

	@Override
	public void normalizeEqual(double p) {
		double norm = 0.0;
		if (Double.isInfinite(p)) {
			for (int i = 0; i < data.length; i++) {
				norm = Math.max(norm, Math.abs(data[i]));
			}
		} else if (p == 1.0) {
			for (int i = 0; i < data.length; i++) {
				norm += Math.abs(data[i]);
			}
		} else if (p == 2.0) {
			for (int i = 0; i < data.length; i++) {
				norm += data[i] * data[i];
			}
			norm = Math.sqrt(norm);
		} else {
			for (int i = 0; i < data.length; i++) {
				norm += Math.pow(Math.abs(data[i]), p);
			}
			norm = Math.pow(norm, 1 / p);
		}
		for (int i = 0; i < data.length; i++) {
			data[i] /= norm;
		}
	}

	@Override
	public VectorIterator iterator() {
		return new DenseVectorIterator();
	}

	private class DenseVectorIterator implements VectorIterator {
		private int cursor = 0;

		@Override
		public boolean hasNext() {
			return cursor < data.length;
		}

		@Override
		public void next() {
			cursor++;
		}

		@Override
		public int getIndex() {
			if (cursor >= data.length) {
				throw new RuntimeException("Iterator out of bound.");
			}
			return cursor;
		}

		@Override
		public double getValue() {
			if (cursor >= data.length) {
				throw new RuntimeException("Iterator out of bound.");
			}
			return data[cursor];
		}
	}
}
